server/api/offline_batch_results.py

```diff
@@ -185,6 +185,89 @@ def record_offline_batch_results(
     return jsonify(status="ok")
 
 
+@api.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/results/batch/",
+    methods=["POST"],
+)
+@restrict_access([UserType.JURISDICTION_ADMIN])
+def record_offline_batch_results_add(
+    election: Election,  # pylint: disable=unused-argument
+    jurisdiction: Jurisdiction,
+    round: Round,
+):
+    offline_batch_results = request.get_json()
+    validate_offline_batch_results(election, jurisdiction, round, offline_batch_results)
+
+    for batch_results in offline_batch_results:
+        for contest_choice_id, result in batch_results["choiceResults"].items():
+            db_session.add(
+                OfflineBatchResult(
+                    jurisdiction_id=jurisdiction.id,
+                    batch_name=batch_results["batchName"],
+                    batch_type=batch_results["batchType"],
+                    contest_choice_id=contest_choice_id,
+                    result=result,
+                )
+            )
+
+    db_session.commit()
+
+    return jsonify(status="ok")
+
+
+@api.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/results/batch/<batch_name>",
+    methods=["PUT"],
+)
+@restrict_access([UserType.JURISDICTION_ADMIN])
+def record_offline_batch_results_update(
+    election: Election,  # pylint: disable=unused-argument
+    jurisdiction: Jurisdiction,
+    round: Round,
+    batch_name: str,
+):
+    offline_batch_results = request.get_json()
+    validate_offline_batch_results(election, jurisdiction, round, offline_batch_results)
+
+    # there should be only one
+    batch_results = offline_batch_results[0]
+
+    if batch_results["batchName"] == batch_name:
+        for contest_choice_id, result in batch_results["choiceResults"].items():
+            batch_result = OfflineBatchResult.query.filter_by(
+                jurisdiction_id=jurisdiction.id,
+                batch_name=batch_results["batchName"],
+                contest_choice_id=contest_choice_id,
+            ).one()
+            batch_result.batch_type = batch_results["batchType"]
+            batch_result.result = result
+            db_session.add(batch_result)
+
+    db_session.commit()
+
+    return jsonify(status="ok")
+
+
+@api.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/results/batch/<batch_name>",
+    methods=["DELETE"],
+)
+@restrict_access([UserType.JURISDICTION_ADMIN])
+def record_offline_batch_results_delete(
+    election: Election,  # pylint: disable=unused-argument
+    jurisdiction: Jurisdiction,
+    round: Round,  # pylint: disable=unused-argument
+    batch_name: str,
+):
+    OfflineBatchResult.query.filter_by(
+        jurisdiction_id=jurisdiction.id, batch_name=batch_name
+    ).delete()
```

Comment:
maybe do a find first to make sure it's actually there, error if not

use case: somebody tries to delete a batch that's already been deleted by somebody else

Reply:
seems okay for the second user to just succeed without an error though?